internal/ssh/internal/mock/server.go

```diff
@@ -0,0 +1,294 @@
+package mock
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+)
+
+type (
+	// server represents an SSH server.
+	//
+	// server is constructed by 'NewServer', can be started (begin listening and
+	// serving connections) by calling its 'ListenAndServe' method. When finished,
+	// a call to 'Shutdown' will gracefully shutdown the TCP listener.
+	server struct {
+		// The SSH server configuration.
+		//
+		// These options may be modified _prior_ to calling 'ListenAndServe',
+		// modifying after will have no effect.
+		Config *ssh.ServerConfig
+
+		// Holds the closure we'll use to shut down the Server.
+		cancel context.CancelFunc
+
+		// The TCP port to listen on.
+		port uint16
+
+		// 'Waiter' is a 'sync.WaitGroup'-like construct, save that it accepts a
+		// 'context.Context' on its 'Done' method, supporting deadlines.
+		wait Waiter
+	}
+	// PubKeyCallback is the function called when the server receives an
+	// authentication attempt via public key. Any non-nil error returned will
+	// immediately abort the connection.
+	PubKeyCallback func(ssh.ConnMetadata, ssh.PublicKey) (*ssh.Permissions, error)
+
+	// ReqChannel produces all *ssh.Requests, which are out-of-band well-known
+	// marshaled data structures which arrive from either a specific channel or
+	// the ssh.SSHConn.
+	ReqChannel <-chan *ssh.Request
+
+	// MsgChannel produces all messages which arrive directly over the ssh
+	// connection (think simple writes to stdin on the client's side).
+	MsgChannel <-chan string
+)
+
+func NewServer(t *testing.T, port uint16, signer ssh.Signer, fn PubKeyCallback) (*server, error) {
+	if t == nil {
+		return nil, fmt.Errorf("no *testing.T provided in call to NewServer")
+	}
+	require.NotNil(t, fn != nil, "a non-nil public key callback is required")
+	require.NotNil(t, signer, "a non-nil ssh.Signer is required")
+	// Init the SSH server config, add the host key
+	config := &ssh.ServerConfig{
+		PublicKeyCallback: fn,
+	}
+	config.AddHostKey(signer)
+	return &server{
+		Config: config,
+
+		wait: NewWaiter(),
+		port: port,
+	}, nil
+}
+
+func (self *server) ListenAndServe(t *testing.T, ctx context.Context) (ReqChannel, MsgChannel, error) {
```

Comment:
The linter doesn't like you naming things `self`, and I don't have the heart to argue with it. 😆